[CLOSED] Light Freshness Edit: ASP.NET and EF Core - 3 tutorials (#37014)
dotnet/AspNetCore.Docs: aspnetcore/data/ef-mvc/crud.md

author: tdykstra
ms.author: tdykstra
ms.custom: mvc
ms.date: 04/16/2026
ms.topic: tutorial
uid: data/ef-mvc/crud

# customer intent: As an ASP.NET developer, I want to make changes to the auto-generated CRUD code, so I can customize my controllers and views.
---

# Tutorial: Implement basic CRUD functionality - ASP.NET MVC with EF Core

In the previous tutorial, you created an MVC application that stores and displays data by using the [Entity Framework (EF) Core](/ef/core/) and a SQL Server local database. In this exercise, you review and customize the CRUD (create, read, update, delete) code that the MVC scaffolding automatically creates for you in controllers and views.

> [!NOTE]
> It's a common practice to implement the repository pattern in order to create an abstraction layer between your controller and the data access layer. To keep the examples simple and focused on demonstrating how to use the Entity Framework itself, the tutorials don't use repositories. For information about repositories with Entity Framework, see [the last tutorial in this series](advanced.md).

In this tutorial, you:

> [!div class="checklist"]

> > * Customize pages: **Details**, **Create**, **Edit**, and **Delete**
> > * Explore how to protect against overposting
> > * Use different approaches for HttpPost `Edit` code and `Delete`
> > * Work with no-tracking queries

Reviewer: ```suggestion
> * Customize the Details page by adding enrollment data with eager loading
> * Update the Create page with security and error handling
> * Update the Edit page to prevent overposting
> * Update the Delete page with error reporting
> * Close database connections
```
For the "In this tutorial, you" section: The verbs in the original are chosen carefully. "Customize" means adding entirely new functionality to the scaffolded code, while "Update" means refining the existing scaffolded code. That said, you hit on a real and important clarity problem to address: Briefly stating why each page is being changed helps the developer quickly decide if this tutorial is relevant to them. I'll make a suggestion here.

Just a side note: if you can use the reviewer's suggested changes as is, it is nice to just hit the "Apply Suggestion" button so their commit gets added.  It gives reviewers credit for collaborating.  If that messes up your process though and need to add it to your own locally as your own commit, that's understandable and no worries.

:::image type="content" source="crud/_static/student-details.png" border="false" alt-text="Screenshot of the Student Details page.":::

## Update the Create page

In the _StudentsController.cs_ file, modify the HttpPost `Create` method by adding a `try-catch` block and removing the `ID` property from the `Bind` attribute.

[!code-csharp[](intro/samples/cu/Controllers/StudentsController.cs?name=snippet_Create&highlight=4,6-7,14-21)]

This code adds the `Student` entity created by the ASP.NET Core MVC model binder to the `Students` entity set, and then saves the changes to the database.

_Model binder_ refers to the ASP.NET Core MVC functionality that makes it easier for you to work with data submitted by a form. A model binder converts posted form values to CLR types and passes them to the action method in parameters. In this case, the model binder instantiates a `Student` entity for you by using property values from the `Form` collection.

You removed the `ID` property from the `Bind` attribute because the ID is the primary key value that SQL Server sets automatically when it inserts the row. Input from the user doesn't set the ID value.

> Other than the `Bind` attribute, the `try-catch` block is the only change made to the scaffolded code. If an exception that derives from the `DbUpdateException` handler is caught while the changes are being saved, a generic error message is displayed. The cause for a `DbUpdateException` exception can be external to the application rather than a programming error, so the user is advised to try again. A production quality application logs such exception, but that functionality isn't implemented in this sample. For more information, see [Logging in .NET and ASP.NET Core](../../fundamentals/logging/index.md).

Reviewer: ```suggestion
Other than the `Bind` attribute, the `try-catch` block is the only change made to the scaffolded code. If an exception that derives from `DbUpdateException` is caught while the changes are being saved, a generic error message is displayed. The cause for a `DbUpdateException` exception can be external to the application rather than a programming error, so the user is advised to try again. A production quality application logs such exception, but that functionality isn't implemented in this sample. For more information, see [Logging in .NET and ASP.NET Core](../../fundamentals/logging/index.md).
```
`DbUpdateException` is an exception class, not a handler. Adding "handler" is technically incorrect. In .NET, "handler" implies something that processes events or exceptions, which is the opposite of what an exception class is.  Given the context of "derived from" is already stated, I wouldn't add "class" instead either.  The original is clear and accurate in this case.

> The `ValidateAntiForgeryToken` attribute helps prevent cross-site request forgery (CSRF) attacks. The [Form Tag Helper](../../mvc/views/working-with-forms.md#the-form-tag-helper) method automatically injects the token into the view. The token is also included when the user submits the form. The `ValidateAntiForgeryToken` attribute checks the token. For more information, see t[Prevent Cross-Site Request Forgery (XSRF/CSRF) attacks in ASP.NET Core](../../security/anti-request-forgery.md) and [FormTagHelper Class](/dotnet/api/microsoft.aspnetcore.mvc.taghelpers.formtaghelper) reference.

Reviewer: ```suggestion
The `ValidateAntiForgeryToken` attribute helps prevent cross-site request forgery (CSRF) attacks. The [Form Tag Helper](../../mvc/views/working-with-forms.md#the-form-tag-helper) automatically injects the token into the view. The token is also included when the user submits the form. The `ValidateAntiForgeryToken` attribute checks the token. For more information, see t[Prevent Cross-Site Request Forgery (XSRF/CSRF) attacks in ASP.NET Core](../../security/anti-request-forgery.md) and [FormTagHelper Class](/dotnet/api/microsoft.aspnetcore.mvc.taghelpers.formtaghelper) reference.
```
`FormTagHelper` is a class (a Tag Helper), not a method. Calling it a "method" would be confusing. Given the context already provided, I think it would be best to remove "method" here and not add "class" in its place either.

> The `ValidateAntiForgeryToken` attribute helps prevent cross-site request forgery (CSRF) attacks. The [Form Tag Helper](../../mvc/views/working-with-forms.md#the-form-tag-helper) method automatically injects the token into the view. The token is also included when the user submits the form. The `ValidateAntiForgeryToken` attribute checks the token. For more information, see t[Prevent Cross-Site Request Forgery (XSRF/CSRF) attacks in ASP.NET Core](../../security/anti-request-forgery.md) and [FormTagHelper Class](/dotnet/api/microsoft.aspnetcore.mvc.taghelpers.formtaghelper) reference.

Reviewer: ```suggestion
The `ValidateAntiForgeryToken` attribute helps prevent cross-site request forgery (CSRF) attacks. The [Form Tag Helper](../../mvc/views/working-with-forms.md#the-form-tag-helper) method automatically injects the token into the view. The token is also included when the user submits the form. The `ValidateAntiForgeryToken` attribute checks the token. For more information, see [Prevent Cross-Site Request Forgery (XSRF/CSRF) attacks in ASP.NET Core](../../security/anti-request-forgery.md) and [FormTagHelper Class](/dotnet/api/microsoft.aspnetcore.mvc.taghelpers.formtaghelper) reference.
```

Removed a typo of "t" in "t[Prevent"

Reviewer: @GitHubber17,  (note this is the old closed PR)  I suspect this was the problem around the error that occured when applying the suggestion. I made two suggested changes on the same line for two problems, without realizing it.  This may have created its own conflict when batching these.  Sorry about that.  I should have combined this into one suggested change.  doh.

An entity can be in one of the following states:

* `Added`: The entity doesn't yet exist in the database. The `SaveChanges` method issues an `INSERT` statement.

* `Unchanged`: Nothing needs to be done with this entity by the `SaveChanges` method. When you read an entity from the database, the entity starts out with this status.

* `Modified`: Some or all of the entity's property values are modified. The `SaveChanges` method issues an `UPDATE` statement.

* `Deleted`: The entity is marked for deletion. The `SaveChanges` method issues a `DELETE` statement.

* `Detached`: The database context doesn't track the entity.

In a desktop application, state changes are typically set automatically. You read an entity and make changes to some of its property values. This behavior causes its entity state to automatically change to `Modified`. When you call the `SaveChanges` method, the Entity Framework generates a SQL `UPDATE` statement that updates only the actual properties that you changed.

> In a web app, the `DbContext` method that initially reads an entity and displays its data to be edited is disposed after a page is rendered. When the HttpPost `Edit` action method is called, a new web request is made and you have a new instance of the `DbContext` method. If you re-read the entity in that new context, you simulate desktop processing.

Reviewer: ```suggestion
In a web app, the `DbContext` that initially reads an entity and displays its data to be edited is disposed after a page is rendered. When the HttpPost `Edit` action method is called, a new web request is made and you have a new instance of the `DbContext`. If you re-read the entity in that new context, you simulate desktop processing.
```

I know this is still a draft, but since I spotted this I'll mention the `DbContext` is not a method, it is an instance of a class.  Since we already describe what it is and how it is used just after, I would not use "instance" either and just drop the "method" added here (and anywhere else it was added).

## Disable tracking of entity objects (no-tracking queries)

When a database context retrieves table rows and creates entity objects that represent them, it tracks whether the entities in memory are in sync with the database, by default. The data in memory acts as a cache and is used when you update an entity. This caching is often unnecessary in a web application because context instances are typically short-lived (a new one is created and disposed for each request). The context that reads an entity is typically disposed before that entity is used again.

You can disable tracking of entity objects in memory by calling the `AsNoTracking` method. Here are some of the common scenarios for this action:

> * During the context lifetime, you don't need to update any entities, and you don't need Entity Framework to [automatically load navigation properties with entities retrieved by separate queries](read-related-data.md). These conditions are frequently met in a controller's HttpGet `Action` methods.

Reviewer: ```suggestion
* During the context lifetime, you don't need to update any entities, and you don't need Entity Framework to [automatically load navigation properties with entities retrieved by separate queries](read-related-data.md). These conditions are frequently met in a controller's HttpGet action methods.
```
Putting `Action` in backticks implies the Action class or attribute in ASP.NET Core. In this case it is referring to controller methods that respond to HTTP requests so keep "action methods" in lowercase, no backticks.
